File: matlab/src/Inference.m
classdef Inference
    % INFERENCE Module 3: Statistical Inference execution.

    properties
        Data
    end

    methods
        function obj = Inference(data)
            obj.Data = data;
        end

        function result = run_ttest(obj, value_col, group_col, paired, equal_var)
            if nargin < 5, equal_var = true; end
            if nargin < 4, paired = false; end

            y = obj.Data.(value_col);
            groups = unique(obj.Data.(group_col));

            if length(groups) ~= 2
                error('T-test requires exactly 2 groups.');
            end

            g1 = y(obj.Data.(group_col) == groups(1));
            g2 = y(obj.Data.(group_col) == groups(2));

            if paired
                [h, p, ci, stats] = ttest(g1, g2);
                name = 'Paired T-test';
            else
                vartype = 'equal';
                if ~equal_var, vartype = 'unequal'; end
                [h, p, ci, stats] = ttest2(g1, g2, 'Vartype', vartype);
                name = 'Independent T-test';
            end

            result = struct('Test', name, 'h', h, 'p', p, 'tstat', stats.tstat, 'df', stats.df);
            disp('--- Inference Result ---');
            disp(result);
        end

        function result = run_mann_whitney(obj, value_col, group_col)
            y = obj.Data.(value_col);
            groups = unique(obj.Data.(group_col));

             if length(groups) ~= 2
                error('Mann-Whitney requires 2 groups.');
            end

            g1 = y(obj.Data.(group_col) == groups(1));
            g2 = y(obj.Data.(group_col) == groups(2));

            [p, h, stats] = ranksum(g1, g2);
            result = struct('Test', 'Mann-Whitney U', 'h', h, 'p', p, 'zval', stats.zval);
            disp('--- Inference Result ---');
            disp(result);
        end

        function result = run_lmm(obj, formula)
            % Runs Linear Mixed Model using fitlme.
            % Formula example: 'Response ~ Condition + (1|SubjectID)'
            result = fitlme(obj.Data, formula);
            disp('--- Inference Result (LMM) ---');
            disp(result);
        end
    end
end

File: python/src/inference.py

import pandas as pd
import scipy.stats as stats
import statsmodels.api as sm
from statsmodels.formula.api import ols, mixedlm

class Inference:
    def __init__(self, data):
        """
        data: pandas DataFrame or dict of arrays/lists.
        """
        if isinstance(data, pd.DataFrame):
            self.data = data
        else:
            self.data = pd.DataFrame(data)

    def run_ttest(self, value_col, group_col, paired=False, equal_var=True):
        """
        Runs T-test (Indep, Welch, or Paired).
        """
        groups = self.data[group_col].unique()
        if len(groups) != 2:
            raise ValueError("T-test requires exactly 2 groups.")

        g1 = self.data[self.data[group_col] == groups[0]][value_col].dropna()
        g2 = self.data[self.data[group_col] == groups[1]][value_col].dropna()

        if paired:
            stat, p = stats.ttest_rel(g1, g2)
            test_name = "Paired T-test"
        else:
            stat, p = stats.ttest_ind(g1, g2, equal_var=equal_var)
            test_name = "Independent T-test" if equal_var else "Welch's T-test"

        return {
            "Test": test_name,
            "Statistic": stat,
            "p-value": p,
            "dof": len(g1) + len(g2) - 2 # Approx
        }

    def run_mann_whitney(self, value_col, group_col):
        """
        Non-parametric independent test.
        """
        groups = self.data[group_col].unique()
        if len(groups) != 2:
            raise ValueError("Mann-Whitney requires 2 groups.")

        g1 = self.data[self.data[group_col] == groups[0]][value_col].dropna()
        g2 = self.data[self.data[group_col] == groups[1]][value_col].dropna()

        stat, p = stats.mannwhitneyu(g1, g2)
        return {"Test": "Mann-Whitney U", "Statistic": stat, "p-value": p}

    def run_lmm(self, formula, group_col):
        """
        Runs Linear Mixed Model.
        formula: string, e.g., 'Response ~ Condition'
        group_col: column name for Random Effects (groups), e.g., 'SubjectID'
        """
        model = mixedlm(formula, self.data, groups=self.data[group_col])
        result = model.fit()
        return result

    def run_anova(self, value_col, group_col):
        """
        One-way ANOVA.
        """
        formula = f'{value_col} ~ C({group_col})'
        model = ols(formula, data=self.data).fit()
        aov_table = sm.stats.anova_lm(model, typ=2)
        return {"Test": "One-way ANOVA", "Table": aov_table}
